pkg/router/template/router.go

```diff
@@ -386,18 +386,20 @@ func (r *templateRouter) readState() error {
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
 func (r *templateRouter) Commit() {
+
 	r.lock.Lock()
-	defer r.lock.Unlock()
 
 	if !r.synced {
 		glog.V(4).Infof("Router state synchronized for the first time")
 		r.synced = true
 		r.stateChanged = true
 	}
 
-	if r.stateChanged {
-		r.rateLimitedCommitFunction.Invoke(r.rateLimitedCommitFunction)
-		r.stateChanged = false
+	needsCommit := r.stateChanged
+	r.lock.Unlock()
+
+	if needsCommit {
+		go r.rateLimitedCommitFunction.Invoke(r.rateLimitedCommitFunction)
```

Comment:
I changed this to call it as a goroutine because this will block under two circumstances:

1. If another thread has the lock pushing (called from another Invoke) which should be fast
2. If another thread has the lock doing the processing of the function call... which could take a while

If the reload is being processed, there is no reason to block the event processing while we wait for that to occur, so we may as well fire off multiple Invokes() since there is no state associated with them and it doesn't matter if a bunch back up.  They will all process quickly and replace one another as soon as the lock is released by the code doing the processing.